Add aim error meter skin component

cdwcgt opened the pull request:
- request from https://github.com/ppy/osu/discussions/25412
- [x] need https://github.com/ppy/osu/pull/27064

Most of the code draws on `BarHitErrorMeter` and `AccuracyHeatmap`

The design is basically derived from mcosu’s position indicator

https://github.com/ppy/osu/assets/34775378/548d8d75-c5ce-4f36-be85-e805ccfcda95

cdwcgt committed:
add HitPositionMeter and basic test

cdwcgt committed:
typo, add border, remove usage of scoreprocessor

cdwcgt committed:
add gameObject to test

cdwcgt committed:
add color for miss

but miss HitEvent have no position?

cdwcgt committed:
adjust object scale

cdwcgt committed:
+ to x

cdwcgt committed:
add settings for hit position meter

cdwcgt committed:
remove empty line

peppy commented on osu.Game.Rulesets.Osu/Skinning/HitPositionMeter.cs:
```diff
@@ -0,0 +1,255 @@
+                Alpha = 0;
+
+                this
+                    .FadeTo(1f, judgement_fade_in_duration, Easing.OutQuint)
```
`FadeInFromZero` is safer here, else you might somehow get a drawable returned to the pool which wasn't at zero alpha causing a sudden transition.

cdwcgt replied:
I am already apply it, but I think it meant [BarHitErrorMeter ](https://github.com/ppy/osu/blob/master/osu.Game/Screens/Play/HUD/HitErrorMeters/BarHitErrorMeter.cs#L479)also need change?

cdwcgt committed:
rename Hit position to Aim error

used in danser and I think this is better

cdwcgt committed:
cleanup

cdwcgt committed:
use FadeInFromZero to avoid sudden transition

cdwcgt committed:
Zoom animation

cdwcgt commented:
![image](https://github.com/ppy/osu/assets/34775378/07377922-cd94-4740-93a3-6b185722495d)
I'm a little undecided. Do we need to add a point in the center as a reference?

RushFTK commented:
I think it's optional and need to do visiual adjust, letting user now it's a not important element. (ref), adding alpha for point itself is fine.

Tanza3D commented:
> ![image](https://private-user-images.githubusercontent.com/34775378/289357845-07377922-cd94-4740-93a3-6b185722495d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDIyMTY1MzAsIm5iZiI6MTcwMjIxNjIzMCwicGF0aCI6Ii8zNDc3NTM3OC8yODkzNTc4NDUtMDczNzc5MjItY2Q5NC00NzQwLTkzYTMtNmIxODU3MjI0OTVkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMTAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjEwVDEzNTAzMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTVhZjg2YzYyNzljY2U4NjY2OTBlZTQ5YjE2ZTkwNzUyOTlkNGUwMTc4MDA3NGQxZWIyOTVlY2Y3Mjg1MTQ1OGUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.cOOTyqiBL5-wPi3xmyM4CMmk5lKe5X0T83AnVr2c1hc) I'm a little undecided. Do we need to add a point in the center as a reference?

perhaps a cross?
![image](https://github.com/ppy/osu/assets/33783503/de047324-795a-4475-b1bc-6c30cd182e16)

Loreos7 commented on osu.Game/Localisation/AimErrorMeterStrings.cs:
```diff
@@ -0,0 +1,64 @@
+        /// <summary>
+        /// "Judgment position size."
+        /// </summary>
+        public static LocalisableString JudgmentSize => new TranslatableString(getKey(@"judgement_line_thickness"), "Judgment position size.");
```
You should avoid dots here to be consistent with the rest UI same elements.
```suggestion
        public static LocalisableString JudgmentSize => new TranslatableString(getKey(@"judgement_line_thickness"), "Judgment position size");
```

longnguyen2004 commented on osu.Game.Rulesets.Osu/Skinning/AimErrorMeter.cs:
Judgment => Judgement for consistency with the rest of the codebase

longnguyen2004 commented on osu.Game/Localisation/AimErrorMeterStrings.cs:
Judgment => Judgement for consistency with the rest of the codebase

cdwcgt committed:
Judgment -> Judgement

cdwcgt committed:
remove dot at the end of label string

peppy commented:
I would have preferred if this matches the design/display as the results screen. It's unclear as to the direction of this. Which is to say it's probably not based on direction and therefore doesn't match the results screen at all, creating confusion and potentially being less useful.

cdwcgt commented:
Maybe we can separate them into two modes.

let it like accuracy heatmap is not too difficult, but in this case, the player can know where the previous object is and most of it is used for playback to check where its clicked.
So I want keep both these display mode.

Joehuu commented on osu.Game/Localisation/AimErrorMeterStrings.cs:
Should be moved inside `HUD`.

cdwcgt committed:
fix getkey isn't match the name

cdwcgt committed:
move `AimErrorMeterStrings` to `HUD`

cdwcgt committed:
clear transforms and returned to pool after `Clear()`

from #25747

cdwcgt committed:
fix `prefix` in `AimErrorMeterStrings` not corrently

cdwcgt committed:
expose `FindRelativeHitPosition` method

cdwcgt committed:
Add new hit position style, change appearance

1. round the hit position that it will not beyond meter range.
2. add relative position style.
- in relative style, rotation can be apply that adjust the relative direction, in Absolute will use `UprightAspectMaintainingContai`ner to prevent rotate because it will cause confusing and meaningless.
3. use the cross-style in https://github.com/ppy/osu/pull/25716#issuecomment-1848974233

cdwcgt committed:
move AimErrorMeter to HUD

cdwcgt committed:
rename param

cdwcgt committed:
Rewrite some text and names

cdwcgt committed:
add comment, fix some

cdwcgt committed:
Merge branch 'master' into hitposition

cdwcgt committed:
Remove dependencies on ScoreProcesser

cdwcgt committed:
Merge branch 'master' into hitposition

cdwcgt committed:
code format

cdwcgt committed:
Merge branch 'master' into hitposition

peppy committed:
Merge branch 'master' into hitposition

cdwcgt committed:
improve testing

can change aim meter style in test

cdwcgt committed:
Merge branch 'master' into hitposition

cdwcgt committed:
handle IApplicableToDifficulty for CS change.

cdwcgt committed:
Merge remote-tracking branch 'upstream/master' into hitposition

peppy commented:
I'm not a huge fan of the diagonal lines in the display, or the general look of the display.

@arflyte did you have any updated visual for this, or any quick thoughts on what's presented in the video in the OP?

cdwcgt committed:
Merge remote-tracking branch 'upstream/master' into hitposition

bdach committed:
Merge branch 'master' into hitposition

bdach committed:
Rename everything to start with

bdach committed:
Improve test coverage somewhat

bdach committed:
Fix manual click test being broken

bdach committed:
Rewrite incomprehensible comments

bdach committed:
Make extracted helper more comprehensible

bdach committed:
Remove undesirable switch syntax

bdach committed:
Fix average marker not moving to first hit position

bdach committed:
Remove unnecessary condition

bdach committed:
Adjust weighted average to 90/10 to match bar error meter

peppy committed:
Adjust marker style description to look more correct in dropdown

peppy commented:
The reset on seek is a bit unfortunate but I think we should revisit that in the future. This works fine otherwise 👍 .
